From 1ffe9385b7c5d9db36e0855785737345bea71e22 Mon Sep 17 00:00:00 2001
From: Edward Savage-Jones <edward.savage-jones@sony.com>
Date: Fri, 13 May 2022 00:49:59 +0200
Subject: [PATCH 1/4] Fix permissions xml and provider access

ThemePicker need to add its use of permissions and declare the
priv app permission parts in an xml file to be able to start.

ThemePicker also need to declare which other apps it intends to
interact with in queries. This enables the Launcher grid/icon preview.

Test: Manual
Bug: 274634330
Change-Id: I55fb94279abeab15dd16a6cabc0bf086c621fc42
---
 Android.bp                        |  9 +++++++++
 AndroidManifest.xml               | 12 ++++++++++++
 android.software.theme_picker.xml | 28 ++++++++++++++++++++++++++++
 3 files changed, 49 insertions(+)
 create mode 100644 android.software.theme_picker.xml

diff --git a/Android.bp b/Android.bp
index f6c85581..7c77533e 100644
--- a/Android.bp
+++ b/Android.bp
@@ -94,6 +94,8 @@ java_defaults {
         ":ThemePicker_src_overrides",
     ],
 
+    required: ["android.software.theme_picker.xml"],
+
     use_embedded_native_libs: true,
 
     resource_zips: [":WallpaperPicker2_res", ":ThemePicker_res", ":ThemePicker_res_overrides"],
@@ -108,6 +110,13 @@ java_defaults {
     system_ext_specific: true,
 }
 
+prebuilt_etc {
+    name: "android.software.theme_picker.xml",
+    system_ext_specific: true,
+    sub_dir: "permissions",
+    src: "android.software.theme_picker.xml",
+}
+
 //
 // Build app code.
 //
diff --git a/AndroidManifest.xml b/AndroidManifest.xml
index 4e71bcc6..2dd11fbb 100755
--- a/AndroidManifest.xml
+++ b/AndroidManifest.xml
@@ -11,8 +11,17 @@
     <uses-permission android:name="android.permission.READ_DEVICE_CONFIG" />
     <uses-permission android:name="android.permission.MODIFY_DAY_NIGHT_MODE" />
     <uses-permission android:name="android.permission.CUSTOMIZE_SYSTEM_UI" />
+    <uses-permission android:name="android.permission.BIND_WALLPAPER"/>
+    <uses-permission android:name="android.permission.READ_MEDIA_IMAGES"/>
+    <uses-permission android:name="android.permission.READ_WALLPAPER_INTERNAL"/>
+    <uses-permission android:name="android.permission.SET_WALLPAPER"/>
 
     <queries>
+        <package android:name="android"/>
+        <package android:name="com.android.launcher3"/>
+        <package android:name="com.android.settings"/>
+        <package android:name="com.android.systemui"/>
+
         <!-- Specific intents Wallpaper picker query for -->
         <!-- Package for theme stub -->
         <package android:name="com.android.customization.themes" />
@@ -45,6 +54,9 @@
         </intent>
     </queries>
 
+    <uses-permission android:name="com.android.launcher3.permission.READ_SETTINGS" />
+    <uses-permission android:name="com.android.launcher3.permission.WRITE_SETTINGS" />
+
     <application
         tools:replace="android:icon,android:name"
         android:extractNativeLibs="false"
diff --git a/android.software.theme_picker.xml b/android.software.theme_picker.xml
new file mode 100644
index 00000000..95d49f40
--- /dev/null
+++ b/android.software.theme_picker.xml
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright (C) 2023 The Android Open Source Project
+
+     Licensed under the Apache License, Version 2.0 (the "License");
+     you may not use this file except in compliance with the License.
+     You may obtain a copy of the License at
+
+          http://www.apache.org/licenses/LICENSE-2.0
+
+     Unless required by applicable law or agreed to in writing, software
+     distributed under the License is distributed on an "AS IS" BASIS,
+     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+     See the License for the specific language governing permissions and
+     limitations under the License.
+-->
+<permissions>
+    <privapp-permissions package="com.android.wallpaper">
+        <permission name="android.permission.BIND_WALLPAPER"/>
+        <permission name="android.permission.CHANGE_OVERLAY_PACKAGES"/>
+        <permission name="android.permission.CUSTOMIZE_SYSTEM_UI"/>
+        <permission name="android.permission.MODIFY_DAY_NIGHT_MODE"/>
+        <permission name="android.permission.READ_DEVICE_CONFIG"/>
+        <permission name="android.permission.READ_WALLPAPER_INTERNAL"/>
+        <permission name="android.permission.SET_WALLPAPER"/>
+        <permission name="android.permission.SET_WALLPAPER_COMPONENT"/>
+        <permission name="android.permission.WRITE_SECURE_SETTINGS"/>
+    </privapp-permissions>
+</permissions>
-- 
2.34.1

